stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,855 @@
+  Eigen::VectorXd abs_tol_f_;
+  Eigen::VectorXd abs_tol_b_;
+  const int lmm_f_;
+  const F f_;
+  const Eigen::Matrix<T_y0_t0, Eigen::Dynamic, 1> y0_;
+  const T_t0 t0_;
+  const std::vector<T_ts> ts_;
+  // std::tuple<T_Args...> args_tuple_;
+  std::tuple<
+      plain_type_t<decltype(deep_copy_vars(std::declval<const T_Args&>()))>...>
+      local_args_tuple_;
+  std::tuple<plain_type_t<decltype(value_of(std::declval<const T_Args&>()))>...>
+      value_of_args_tuple_;
+  std::vector<Eigen::VectorXd> y_;
+  void* cvodes_mem_;
+  Eigen::VectorXd state;
+  N_Vector nv_state_;
+  N_Vector nv_abs_tol_f_;
+  N_Vector nv_abs_tol_b_;
+  SUNMatrix A_f_;
+  SUNLinearSolver LS_f_;
+
+  template <require_eigen_col_vector_t<T_y0>* = nullptr>
+  cvodes_integrator_adjoint_memory(Eigen::VectorXd abs_tol_f,
+                                   Eigen::VectorXd abs_tol_b, int lmm_f,
+                                   const F& f, const T_y0& y0, const T_t0& t0,
+                                   const std::vector<T_ts>& ts,
+                                   const T_Args&... args)
+      : N_(y0.size()),
+        abs_tol_f_(abs_tol_f),
+        abs_tol_b_(abs_tol_b),
+        lmm_f_(lmm_f),
+        f_(f),
+        y0_(y0),
+        t0_(t0),
+        ts_(ts),
+        // args_tuple_(std::make_tuple(args...)),
+        local_args_tuple_(deep_copy_vars(args)...),
+        value_of_args_tuple_(value_of(args)...),
+        y_(ts_.size()),
+        cvodes_mem_(nullptr),
+        state(value_of(y0)) {
+    if (N_ > 0) {
+      nv_state_ = N_VMake_Serial(N_, state.data());
+      nv_abs_tol_f_ = N_VMake_Serial(N_, &abs_tol_f_(0));
+      nv_abs_tol_b_ = N_VMake_Serial(N_, &abs_tol_b_(0));
+      A_f_ = SUNDenseMatrix(N_, N_);
+      LS_f_ = SUNDenseLinearSolver(nv_state_, A_f_);
+
+      cvodes_mem_ = CVodeCreate(lmm_f_);
+      if (cvodes_mem_ == nullptr) {
+        throw std::runtime_error("CVodeCreate failed to allocate memory");
+      }
+    }
+  }
+
+  ~cvodes_integrator_adjoint_memory() {
+    if (N_ > 0) {
+      SUNLinSolFree(LS_f_);
+      SUNMatDestroy(A_f_);
+
+      N_VDestroy_Serial(nv_state_);
+      N_VDestroy_Serial(nv_abs_tol_f_);
+
+      if (cvodes_mem_) {
+        CVodeFree(&cvodes_mem_);
+      }
+    }
+  }
+
+  template <typename yT, typename... ArgsT>
+  constexpr auto rhs(double t, const yT& y, std::ostream* msgs,
+                     const std::tuple<ArgsT...>& args_tuple) const {
+    return apply([&](auto&&... args) { return f_(t, y, msgs, args...); },
+                 args_tuple);
+  }
+
+  friend class cvodes_integrator_adjoint_vari<F, T_y0, T_t0, T_ts, T_Args...>;
+};
+
+/**
+ * Integrator interface for CVODES' ODE solvers (Adams & BDF
+ * methods).
+ *
+ * @tparam F Type of ODE right hand side
+ * @tparam T_y0 Type of scalars for initial state
+ * @tparam T_param Type of scalars for parameters
+ * @tparam T_t0 Type of scalar of initial time point
+ * @tparam T_ts Type of time-points where ODE solution is returned
```

Comment:
Are all these docs correct?

Reply:
`@tparam T_param Type of scalars for parameters` is wrong, the rest is fine. PArameter types are now part of the tuple thing.